unsloth_zoo/compiler.py

```diff
@@ -519,120 +517,120 @@ def create_new_function(
     try: trl_version = importlib_version("trl")
     except: trl_version = "0"
 
-    versioning = '"""\n' + \
+    versions = '"""\n' + \
         f'{unsloth_zoo_version}\n'\
         f'{unsloth_version}\n'\
         f'{transformers_version}\n'\
         f'{trl_version}\n__UNSLOTH_VERSIONING__\n' + '"""\n'
 
-    write_new_source = versioning + new_source
+    new_source = versions + new_source
 
     # Write function
     global UNSLOTH_COMPILE_USE_TEMP
     file_source = None
-    compile_folder, UNSLOTH_COMPILE_USE_TEMP = get_compile_folder(use_tempfile = False)
-    function_location = os.path.join(compile_folder, f"{name}.py")
+    compile_dir, UNSLOTH_COMPILE_USE_TEMP = get_compile_folder(use_tempfile = False)
+    function_file = os.path.join(compile_dir, f"{name}.py")
 
     # Check if file was already created!
-    if not overwrite and os.path.isfile(function_location):
+    lock_file = function_file + ".lock"
 
-        # Check if exactly equivalent
-        with open(function_location, "r", encoding = "utf-8") as f:
-            file_source = f.read()
+    with FileLock(lock_file):
+        if not overwrite and os.path.isfile(function_file):
+            # Check if exactly equivalent
+            with open(function_file, "r", encoding = "utf-8") as f:
+                file_source = f.read()
 
-        if file_source != write_new_source:
-            overwrite = True
-        elif not overwrite:
-            if "__UNSLOTH_VERSIONING__" not in file_source:
+            if file_source != new_source:
                 overwrite = True
-            else:
-                versions = file_source[:file_source.find('__UNSLOTH_VERSIONING__')]
-                if versioning[:versioning.find('__UNSLOTH_VERSIONING__')] != versions:
+            elif not overwrite:
```

Comment:
BTW, can this be removed? If the above is false this seems like it will always be false.